rag/llm/chat_model.py

```
@@ -549,6 +549,23 @@ async def _exec_tool(tc):
                     logging.info(f"Response tool_calls={response.choices[0].message.tool_calls}")
                     results = await asyncio.gather(*[_exec_tool(tc) for tc in response.choices[0].message.tool_calls])
+                    # Terminal-tool short-circuit (mirror of the streaming
+                    # variant at the top of this file): a terminal tool already
+                    # produces the final answer, so return its result instead of
+                    # feeding it back for another LLM round. Without this the
+                    # non-streaming react loop keeps re-invoking `rag` every
+                    # round (Q654 spun 17 tree passes → 15 min). `rag` is always
+                    # terminal, so default to {"rag"} even if a probe wrapper
+                    # dropped the configured terminal_tools.
+                    _terminal = getattr(self, "terminal_tools", None) or {"rag"}
+                    for tc, name, args, result, err in results:
```

Ruff (0.16.2)

[warning] Loop control variable `tc` not used within loop body

Rename unused `tc` to `_tc`

(B007)